### PR TITLE
Renaming of revlex, negrevlex, introduction of deginvlex

### PR DESCRIPTION
**Pull request by ederc**
This PR was initiated by https://github.com/oscar-system/Oscar.jl/issues/2950.

  1. Renames revlex to invlex and negrevlex to neginvlex.
  2. Introduces the monomial ordering deginvlex.
  3. Keeps (until next breaking release) revlex and negrevlex available (added to deprecations list)
  4. Bumps `Singular.jl` to version `0.20.0`.

### CODE REVIEW

File: src/Rings/orderings.jl [hunk 1]
```diff
@@ -453,32 +456,34 @@ Given a vector `V` of variables, return the reverse lexicographical ordering on
...
 
-function revlex(v::AbstractVector{<:MPolyRingElem})
+function invlex(v::AbstractVector{<:MPolyRingElem})
   i = _unique_var_indices(v)
-  return MonomialOrdering(parent(first(v)), SymbOrdering(:revlex, i))
+  return MonomialOrdering(parent(first(v)), SymbOrdering(:invlex, i))
 end
 
-function _matrix(nvars::Int, o::SymbOrdering{:revlex})
+const revlex = invlex
```
Comment (fingolfin) [inline src/Rings/orderings.jl]: Why is this in here? Wasn't the plan to get rid of this meaning of `revlex` (either removing `revlex`, or giving it a new meaning -- though I'd still vote for first removing it, at least for a time, so that code using it can be caught and adapted)
```suggestion
```

File: src/exports.jl [hunk 2]
```diff
@@ -737,6 +738,8 @@ export inverse
 export invert
 export invert_birational_map
 export inverted_set
+export invlex
+export revlex # to be removed
```
Comment (fingolfin) [inline src/exports.jl]: This export list is meant to be sorted alphabetically. And anyway, I think `@deprecate` exports the old name by default, so this line should just not be needed, I think...
```suggestion
```

File: src/Rings/orderings.jl [hunk 3]
```diff
@@ -25,12 +28,12 @@ export monomial_ordering
 export negdeglex
 export negdegrevlex
 export neglex
+export neginvlex
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
export neginvlex
export neglex
```

File: src/Rings/orderings.jl [hunk 4]
```diff
@@ -629,6 +698,8 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
   return 0
 end
 
+const negrevlex = neginvlex
```
Comment (fingolfin) [inline src/Rings/orderings.jl]: Why is this needed?

File: src/Rings/orderings.jl [hunk 5]
```diff
@@ -488,7 +493,71 @@ function _matrix(nvars::Int, o::SymbOrdering{:revlex})
...
+  end
+  return 0
+end
+
+#### deginvlex ####
+
+@doc raw"""
+    deginvlex(R::MPolyRing) -> MonomialOrdering
+
+Return the reverse lexicographical ordering on the set of monomials in the variables of `R`.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Return the degree inverse lexicographical ordering on the set of monomials in the variables of `R`.
```

File: docs/src/CommutativeAlgebra/GroebnerBases/orderings.md [hunk 6]
```diff
@@ -159,14 +159,24 @@ $x^\alpha > x^\beta \;  \Leftrightarrow \;  \deg(x^\alpha) > \deg(x^\beta)  \;\t
...
 ```@docs
-revlex(R::MPolyRing)
+invlex(R::MPolyRing)
+```
+
+#### The Degree Inverse Lexicographical Ordering
+
+The *degree inverse lexicographical ordering* `deginvlex` is defined by setting
+
+$x^\alpha > x^\beta \;  \Leftrightarrow \deg(x^\alpha) > \deg(x^\beta)  \;\text{ or } \;\exists \; 1 \leq i \leq n: \alpha_n = \beta_n, \dots, \alpha_{i+1} = \beta_{i+1}, \alpha_i  > \beta_i.$
```
Comment (fingolfin) [inline docs/src/CommutativeAlgebra/GroebnerBases/orderings.md]: Not a change request for this PR, but perhaps for a future round of improvements:

I think we should move all these definitions into the docstrings, so that they are also available in the REPL. (Alternatively: leave these nicely LaTeX formatted defs here, and use a simplified version better suited for viewing in the REPL in the docstrings).

File: src/Rings/orderings.jl [hunk 7]
```diff
@@ -437,14 +440,14 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
...
 end
 
-#### revlex ####
+#### invlex ####
 
 @doc raw"""
-    revlex(R::MPolyRing) -> MonomialOrdering
+    invlex(R::MPolyRing) -> MonomialOrdering
 
 Return the reverse lexicographical ordering on the set of monomials in the variables of `R`.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Return the inverse lexicographical ordering on the set of monomials in the variables of `R`.
```

File: src/Rings/orderings.jl [hunk 8]
```diff
@@ -437,14 +440,14 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
...
 
-    revlex(V::AbstractVector{<:MPolyRingElem}) -> MonomialOrdering
+    invlex(V::AbstractVector{<:MPolyRingElem}) -> MonomialOrdering
 
 Given a vector `V` of variables, return the reverse lexicographical ordering on the set of monomials in these variables.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Given a vector `V` of variables, return the inverse lexicographical ordering on the set of monomials in these variables.
```

File: src/Rings/orderings.jl [hunk 9]
```diff
@@ -488,7 +493,71 @@ function _matrix(nvars::Int, o::SymbOrdering{:revlex})
...
+#### deginvlex ####
+
+@doc raw"""
+    deginvlex(R::MPolyRing) -> MonomialOrdering
+
+Return the reverse lexicographical ordering on the set of monomials in the variables of `R`.
+
+    deginvlex(V::AbstractVector{<:MPolyRingElem}) -> MonomialOrdering
+
+Given a vector `V` of variables, return the reverse lexicographical ordering on the set of monomials in these variables.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Given a vector `V` of variables, return the degree inverse lexicographical ordering on the set of monomials in these variables.
```

File: src/Rings/orderings.jl [hunk 10]
```diff
@@ -565,14 +634,14 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
...
 end
 
-#### negrevlex ####
+#### neginvlex ####
 
 @doc raw"""
-    negrevlex(R::MPolyRing) -> MonomialOrdering
+    neginvlex(R::MPolyRing) -> MonomialOrdering
 
 Return the negative reverse lexicographical ordering  on the set of monomials in the variables of `R`.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Return the negative inverse lexicographical ordering on the set of monomials in the variables of `R`.
```

File: src/Rings/orderings.jl [hunk 11]
```diff
@@ -565,14 +634,14 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
...
 
-    negrevlex(V::AbstractVector{<:MPolyRingElem}) -> MonomialOrdering
+    neginvlex(V::AbstractVector{<:MPolyRingElem}) -> MonomialOrdering
 
 Given a vector `V` of variables, return the negative reverse lexicographical ordering on the set of monomials in these variables.
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
Given a vector `V` of variables, return the negative inverse lexicographical ordering on the set of monomials in these variables.
```

File: src/Rings/orderings.jl [hunk 12]
```diff
@@ -1922,10 +1993,13 @@ function _try_singular_easy(Q::order_conversion_ctx, o::SymbOrdering{S}) where S
...
   return S == :lex          ? (true, Singular.ordering_lp(n)) :
          S == :revlex       ? (true, Singular.ordering_rp(n)) :
+         S == :invlex       ? (true, Singular.ordering_ip(n)) :
+         S == :deginvlex    ? (true, Singular.ordering_Ip(n)) :
          S == :deglex       ? (true, Singular.ordering_Dp(n)) :
          S == :degrevlex    ? (true, Singular.ordering_dp(n)) :
+         S == :deginvlex    ? (true, Singular.ordering_Ip(n)) :
          S == :neglex       ? (true, Singular.ordering_ls(n)) :
-         S == :negrevlex    ? (true, Singular.ordering_rs(n)) :
+         S == :neginvlex    ? (true, Singular.ordering_rs(n)) :
```
Comment (fingolfin) [inline src/Rings/orderings.jl]:
```suggestion
         S == :neginvlex    ? (true, Singular.ordering_is(n)) :
```

File: src/exports.jl [hunk 13]
```diff
@@ -1065,7 +1068,8 @@ export negbias
 export negdeglex
 export negdegrevlex
 export neglex
```
Comment (fingolfin) [inline src/exports.jl]:
```suggestion
export neginvlex
export neglex
```

File: src/exports.jl [hunk 14]
```diff
@@ -1065,7 +1068,8 @@ export negbias
...
-export negrevlex
+export neginvlex
+export negrevlex # to be removed
```
Comment (fingolfin) [inline src/exports.jl]:
```suggestion
```
neginvlex needs to be moved up one line, and `negrevlex` should be exported by `@deprecate` already

File: src/Rings/orderings.jl (hunk 4 above)
Comment (ederc) [reply src/Rings/orderings.jl]: Has been removed, I just did not understand how the `deprecate` business works.

File: docs/src/CommutativeAlgebra/GroebnerBases/orderings.md (hunk 6 above)
Comment (ederc) [reply docs/src/CommutativeAlgebra/GroebnerBases/orderings.md]: Will discuss this with @wdecker  and handle it in a different PR.

Comment (lgoettgens) [inline docs/src/CommutativeAlgebra/GroebnerBases/orderings.md]: This and all other definitions of `deg` or weighted orderings are wrong due to the "or". There is always an `\deg(x^\alpha) = \deg(x^\beta) and` missing on the right-hand side of the "or".

Comment (ederc) [reply docs/src/CommutativeAlgebra/GroebnerBases/orderings.md]: Should be resolved now.

**Review by fingolfin — APPROVED** [review]
Looks good to me, thanks! But perhaps @wdecker and @hannes14 also would like to comment / approve?

File: src/Rings/orderings.jl [hunk 15]
```diff
@@ -437,48 +438,113 @@ function _cmp_monomials(f::MPolyRingElem, k::Int, g::MPolyRingElem, l::Int, o::S
...
+
+# Examples
+```jldoctest
+julia> R, (w, x, y, z) = polynomial_ring(QQ, ["w", "x", "y", "z"])
+(Multivariate polynomial ring in 4 variables over QQ, QQMPolyRingElem[w, x, y, z])
+
+julia> o1 = deginvlex(R)
+deginvlex([w, x, y, z])
+
+julia> canonical_matrix(o1)
```
Comment (hannes14) [inline src/Rings/orderings.jl:525]: This is the matrix for invlex, not deginvlex

File: src/deprecations.jl [hunk 16]
```diff
@@ -488,3 +488,9 @@ function blow_up(v::NormalToricVarietyType, I::MPolyIdeal, coordinate_name::Stri
   Base.depwarn("The keyword argument set_attributes is deprecated. Please use the function without the keyword.", :blow_up)
   return blow_up(v, I, coordinate_name)
 end
+
+# Deprecated after 0.14.0
+@deprecate revlex(R::MPolyRing) invlex(R::MPolyRing)
+@deprecate revlex(v::AbstractVector{<:MPolyRingElem}) invlex(v::AbstractVector{<:MPolyRingElem})
+@deprecate negrevlex(R::MPolyRing) ngeinvlex(R::MPolyRing)
+@deprecate negrevlex(v::AbstractVector{<:MPolyRingElem}) neginvlex(v::AbstractVector{<:MPolyRingElem})
```
Comment (hannes14) [inline src/deprecations.jl:496]: why deprecate revlex/negrevlex: it is not exported?

Comment (lgoettgens) [reply src/deprecations.jl:496]: they are exported on the current master. `@deprecate` exports them itself, if the second argument is exported as well (so `invlex/neginvlex` here)

File: src/Rings/orderings.jl (hunk 15 above)
Comment (ederc) [reply src/Rings/orderings.jl:525]: Should be fixed now, also the monomial comparison for `deginvlex`.
